cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/IntraBrokerGoalViolationDetector.java

```diff
@@ -0,0 +1,205 @@
+import com.codahale.metrics.Timer;
+import com.linkedin.cruisecontrol.detector.Anomaly;
+import com.linkedin.cruisecontrol.exception.NotEnoughValidWindowsException;
+import com.linkedin.kafka.cruisecontrol.KafkaCruiseControl;
+import com.linkedin.kafka.cruisecontrol.analyzer.AnalyzerUtils;
+import com.linkedin.kafka.cruisecontrol.analyzer.OptimizationOptions;
+import com.linkedin.kafka.cruisecontrol.analyzer.ProvisionResponse;
+import com.linkedin.kafka.cruisecontrol.analyzer.ProvisionStatus;
+import com.linkedin.kafka.cruisecontrol.analyzer.goals.Goal;
+import com.linkedin.kafka.cruisecontrol.async.progress.OperationProgress;
+import com.linkedin.kafka.cruisecontrol.config.KafkaCruiseControlConfig;
+import com.linkedin.kafka.cruisecontrol.config.constants.AnomalyDetectorConfig;
+import com.linkedin.kafka.cruisecontrol.exception.KafkaCruiseControlException;
+import com.linkedin.kafka.cruisecontrol.exception.OptimizationFailureException;
+import com.linkedin.kafka.cruisecontrol.executor.ExecutorState;
+import com.linkedin.kafka.cruisecontrol.model.ClusterModel;
+import com.linkedin.kafka.cruisecontrol.model.ReplicaPlacementInfo;
+import org.apache.kafka.common.TopicPartition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+
+import static com.linkedin.kafka.cruisecontrol.KafkaCruiseControlUtils.ANOMALY_DETECTOR_SENSOR;
+import static com.linkedin.kafka.cruisecontrol.detector.AnomalyDetectorUtils.ANOMALY_DETECTION_TIME_MS_OBJECT_CONFIG;
+import static com.linkedin.kafka.cruisecontrol.detector.AnomalyDetectorUtils.KAFKA_CRUISE_CONTROL_OBJECT_CONFIG;
+
+
+/**
+ * This class will be scheduled to run periodically to check if the given goals are violated or not. An alert will be
+ * triggered if one of the goals is not met.
+ */
+public class IntraBrokerGoalViolationDetector extends BaseGoalViolationDetector implements Runnable {
+  private static final Logger LOG = LoggerFactory.getLogger(IntraBrokerGoalViolationDetector.class);
+  private final List<Goal> _detectionGoals;
+  private final Timer _goalViolationDetectionTimer;
+  private final boolean _populateReplicaInfo = true;
+
+  public IntraBrokerGoalViolationDetector(Queue<Anomaly> anomalies, KafkaCruiseControl kafkaCruiseControl, MetricRegistry dropwizardMetricRegistry) {
+    super(anomalies, kafkaCruiseControl, dropwizardMetricRegistry);
+    KafkaCruiseControlConfig config = _kafkaCruiseControl.config();
+    // Notice that we use a separate set of Goal instances for anomaly detector to avoid interference.
+    _detectionGoals = config.getConfiguredInstances(AnomalyDetectorConfig.ANOMALY_DETECTION_INTRA_BROKER_GOALS_CONFIG, Goal.class);
+    _goalViolationDetectionTimer = dropwizardMetricRegistry.timer(MetricRegistry.name(ANOMALY_DETECTOR_SENSOR,
+            "intra-broker-goal-violation-detection-timer"));
+
+  }
+
+  @Override
+  public void run() {
```

Comment:
This method looks like highly duplicated with the GoalViolationDetector.  I think we should avoid it. E.g. maybe extract the common logic into the parent class.